src/server/src/main/java/io/cassandrareaper/service/RepairManager.java

```diff
@@ -234,65 +234,63 @@ private void abortSegmentsWithNoLeader(RepairRun repairRun, Collection<RepairSeg
             .collect(Collectors.toSet());
 
         LOG.debug("No leader on the following segments : {}", orphanedSegments);
-        abortSegments(orphanedSegments, repairRun, false, true);
+        abortSegments(orphanedSegments, repairRun);
       }
     } finally {
       repairRunnersLock.unlock();
     }
   }
 
-  public RepairSegment abortSegment(UUID repairRunId, UUID segmentId) {
-    RepairSegment segment = context.storage.getRepairSegment(repairRunId, segmentId).get();
-    RepairRun repairRun = context.storage.getRepairRun(repairRunId).get();
-    if (context.storage instanceof IDistributedStorage) {
-      ((IDistributedStorage) context.storage).forceReleaseLead(segmentId);
-      ((IDistributedStorage) context.storage).takeLead(segmentId);
-    }
-    if (null == segment.getCoordinatorHost() || RepairSegment.State.DONE == segment.getState()) {
-      SegmentRunner.postponeSegment(context, segment);
-    } else {
-      abortSegments(Arrays.asList(segment), repairRun, true, false);
+  public RepairSegment abortSegment(UUID runId, UUID segmentId) throws ReaperException {
+    RepairSegment segment = context.storage.getRepairSegment(runId, segmentId).get();
+    try {
+      if (null == segment.getCoordinatorHost() || RepairSegment.State.DONE == segment.getState()) {
+        RepairUnit repairUnit = context.storage.getRepairUnit(segment.getRepairUnitId());
+        UUID leaderElectionId = repairUnit.getIncrementalRepair() ? runId : segmentId;
+        boolean tookLead;
+        if (tookLead = takeLead(context, leaderElectionId) || renewLead(context, leaderElectionId)) {
+          try {
+            SegmentRunner.postponeSegment(context, segment);
+          } finally {
+            if (tookLead) {
+              releaseLead(context, leaderElectionId);
+            }
+          }
+        }
+      } else {
+        abortSegments(Arrays.asList(segment), context.storage.getRepairRun(runId).get());
+      }
+      return context.storage.getRepairSegment(runId, segmentId).get();
+    } catch (AssertionError error) {
+      throw new ReaperException("lead is already taken on " + runId + ":" + segmentId, new Exception(error));
     }
-
-    return context.storage.getRepairSegment(repairRunId, segmentId).get();
   }
 
   void abortSegments(Collection<RepairSegment> runningSegments, RepairRun repairRun) {
```

Comment:
You removed the forced abort feature it seems, which is valuable in the case we have stuck segments in the UI (prevents the requirement to have to modify the storage directly through SQL/CQL).
What's the motivation here?

Reply:
it was only used by one caller, and could be solved with the `tookLead` boolean and by simplifying how  that one called `abortSegments(..)` was forcibly first taking the lead: https://github.com/thelastpickle/cassandra-reaper/blob/master/src/server/src/main/java/io/cassandrareaper/service/RepairManager.java#L254

(it was not like anywhere in the code was actually aborting the segment without holding the lock to begin with. but now the taking and releasing of the lock is better paired and controlled.)